File: sdks/python/taskito/dashboard/_testing.py
"""Shared helpers for dashboard endpoint tests.

The dashboard requires a logged-in session for every API route once setup
is complete. :class:`AuthedClient` wraps the stdlib ``urllib.request`` so
tests can issue authenticated HTTP calls without repeating the cookie /
CSRF dance.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any

from taskito import Queue
from taskito.dashboard.auth import AuthStore, Session


@dataclass(frozen=True)
class AuthedClient:
    """Stateless HTTP helper that attaches session + CSRF to every call."""

    base: str
    session: Session

    @property
    def _cookies(self) -> dict[str, str]:
        return {
            "taskito_session": self.session.token,
            "taskito_csrf": self.session.csrf_token,
        }

    def _cookie_header(self) -> str:
        return "; ".join(f"{k}={v}" for k, v in self._cookies.items())

    def get(self, path: str, *, raise_for_status: bool = True) -> Any:
        url = self.base + path
        req = urllib.request.Request(url, method="GET")
        req.add_header("Cookie", self._cookie_header())
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read() or b"{}")
        except urllib.error.HTTPError as e:
            if raise_for_status:
                raise
            return {"status": e.code, "body": json.loads(e.read() or b"{}")}

    def post(self, path: str, body: dict | None = None) -> Any:
        return self._mutate("POST", path, body)

    def put(self, path: str, body: dict | None = None) -> Any:
        return self._mutate("PUT", path, body)

    def delete(self, path: str) -> Any:
        return self._mutate("DELETE", path, None)

    def _mutate(self, method: str, path: str, body: dict | None) -> Any:
        url = self.base + path
        data = json.dumps(body).encode() if body is not None else b""
        req = urllib.request.Request(url, method=method, data=data)
        req.add_header("Cookie", self._cookie_header())
        req.add_header("X-CSRF-Token", self.session.csrf_token)
        if body is not None:
            req.add_header("Content-Type", "application/json")
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read() or b"{}")


def seed_admin_and_session(
    queue: Queue,
    *,
    username: str = "test-admin",
    password: str = "test-pass-1234",
) -> Session:
    """Create a one-off admin and return a fresh session for it."""
    store = AuthStore(queue)
    if store.get_user(username) is None:
        store.create_user(username, password, role="admin")
    user = store.get_user(username)
    assert user is not None
    return store.create_session(user)

File: sdks/node/src/dashboard/testing.ts
// Test seam for session-authenticated dashboard requests. Internal — used by
// the SDK's own tests; not exported from the package barrel.

import type { Queue } from "../index";
import { AuthStore, type DashboardSession } from "./authStore";

export interface SeededAuth {
  session: DashboardSession;
  /** Headers carrying the session cookie + CSRF pair for state-changing calls. */
  headers: Record<string, string>;
}

/** Create an admin (or given-role) user plus a live session for tests. */
export async function seedAdminAndSession(
  queue: Queue,
  options: { username?: string; password?: string; role?: string } = {},
): Promise<SeededAuth> {
  const store = new AuthStore(queue);
  const user = await store.createUser(
    options.username ?? "admin",
    options.password ?? "password123",
    options.role ?? "admin",
  );
  const session = store.createSession(user);
  return { session, headers: authedHeaders(session) };
}

/** Cookie + CSRF headers for an existing session. */
export function authedHeaders(session: DashboardSession): Record<string, string> {
  return {
    cookie: `taskito_session=${session.token}; taskito_csrf=${session.csrfToken}`,
    "x-csrf-token": session.csrfToken,
  };
}
